docs/proposals/super-simple-discovery-api.md

```diff
@@ -0,0 +1,231 @@
+# Super Simple Discovery API
+
+## Overview
+
+It is surprisingly hard to figure out how to talk to a Kubernetes cluster.  Not only do client need to know where to look on the network, they also need to identify the set of root certificates to trust when talking to that endpoint.  
+
+This presents a set of problems:
+* It should be super easy for users to configure client systems with a minimum of effort `kubectl` or `kubeadm init` (or other client systems).
+  * Establishing this should be doable even in the face of nodes and master components booting out of order.
+  * We should have mechanisms that don't require users to ever have to manually manage certificate files.
+* Over the life of the cluster this information could change and client systems should be able to adapt.
+
+## ClusterInfo object
+
+First we define a set of information that identifies a cluster and how to talk to it.  
+
+This includes:
+
+* A cluster serial number/identifier.  
+  * In an HA world, API servers may come and go and it is necessary to make sure we are talking to the same cluster as we thought we were talking to.
+  * It looks like Federation already has this concept.  We need to back port that to non-federated clusters.
+* A set of addresses for finding the cluster.  
+  * It is implied that all of these are equivalent and that a client can try multiple until an appropriate target is found.
+  * Initially I’m proposing a flat set here.  In the future we can introduce more structure that hints to the user which addresses to try first.
+* A set of root certificates to trust for this cluster.
+  * We make this a set so that root certificates can be rotated.
+  * These are BASE64 encoded PEM values (without `-----BEGIN CERTIFICATE-----` lines).
+  This is similar to how [JWS encodes certifcates in JSON](https://www.google.com/url?q=https://tools.ietf.org/html/rfc7515%23appendix-B&sa=D&ust=1471374807953000&usg=AFQjCNFQNcmbquq3XEegKc76v5j956aSPw).
+* The date and time this information was fetched.
+  * This is so that a client can figure out if this information is out of date.
+  * Perhaps include a TTL?
+
+  Any client of the cluster will want to have this information.  As the configuration of the cluster changes we need the client to keep this information up to date.  It is assumed that the information here won’t drift so fast that clients won’t be able to find *some* way to connect.
+
+  In exceptional circumstances, it is possible that this information may be out of date and a client would be unable to connect to a cluster.  If a user has kubectl set up and working well pointing to raw IP addresses and doesn’t run it for a long time it is possible that both (a) the set of servers will have migrated so that no IP is good or (b) the root certificates will have rotated so that the user can no longer trust the endpoint.
+
+  We serialize this into a JSON object (exact syntax still TBD -- this is a sketch).  The whitespace is inserted for readability.
+  ```json
```

Comment:
Yes.  But `.kube/config` is kind of a hairball that wasn't built to be an API object.  See #https://github.com/kubernetes/kubernetes/issues/30395 for efforts to evolve that.  This certainly plays in.